onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2487,187 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+  auto size_of_output = static_cast<int>(output_shape.size());
+  if (size_of_output > 0) { // try full shape inference
+    // a new dim is always added to the output
+    final_output_shape->add_dim()->set_dim_value(1);
+
+    // input batch dim
+    if (static_cast<size_t>(input_shape.dim_size()) > 2) {
+      *final_output_shape->add_dim() = input_shape.dim(0);
+    }
+
+    // output_shape is appended to the end
+    for (int i = 0; i < size_of_output; ++i) {
+      final_output_shape->add_dim()->set_dim_value(output_shape[i]);
+    }
+  } else { // fallback to rank inference otherwise
+    // a new dim is always added to the output
+    final_output_shape->add_dim();
+
+    // input batch dim
+    if (static_cast<size_t>(input_shape.dim_size()) > 2) {
+      final_output_shape->add_dim();
+    }
+
+    // output_shape is appended to the end
+    int n_output_shape = ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value();
+    for (int i=0; i < n_output_shape; ++i) {
+      final_output_shape->add_dim();
+    }
+  }
+
+  return;
+}
+
+static const char* Col2Im_ver16_doc = R"DOC(
+The operator rearranges the input matrix columns into blocks.
+
+Col2Im behaves like PyTorch's fold for two output spatial dimensions https://pytorch.org/docs/stable/generated/torch.nn.Fold.html
+
+Note that only 4-D output tensors (batched image-like tensors) are supported.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Col2Im,
+    16,
+    OpSchema()
+    .Attr(
+        "dilation",
+        "Dilation value along each spatial axis of the kernel. "
+        "`dilation` format should be a 2-tuple as follow (dim_1, dim_2, ...). "
+        "If not present, the dilation defaults to 1 for each spatial axis",
+        AttributeProto::INTS,
+        OPTIONAL_VALUE)
+	  .Attr(
+        "padding",
+        "Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0. "
+        "The value represent the number of pixels added to the beginning and end part of the corresponding axis. "
+        "`padding` format should be a 2-tuple as follow (begin, end). "
+        "If not present, the padding defaults to 0 along start and end of each spatial axis.",
+        AttributeProto::INTS,
+        OPTIONAL_VALUE)
+    .Attr(
+        "stride",
+        "Stride along each spatial axis. "
+        "`stride` format should be a 2-tuple as follow (dim_1, dim_2, ...). "
+        "If not present, the stride defaults to 1 for each spatial axis.",
+        AttributeProto::INTS,
+        OPTIONAL_VALUE)
+    .SetDoc(Col2Im_ver16_doc)
+    .Input(
+        0,
+        "input",
+        "Input data tensor to be rearranged into blocks. Either (H, W) or (N, H, W).",
+        "T",
+        OpSchema::Single,
+        true,
+        1,
+        OpSchema::Differentiable)
+    .Input(
+        1,
+        "output_shape",
+        "The shape of the input before applying im2col transformation. Either (1, H, W) or (1, N, H, W).",
```

Comment:
Don't understand this. May be one issue is a typo "input vs. output". But even if we take this to be output_shape, it is not clear to me.

Reply:
This is really output_shape. Although it looks redundant, it is needed when dilation > 1. Will add a note justifying this

Reply:
To clarify, I was making two points. First point is the phrase "The shape of the input before" doesn't match the name "output_shape". May be "The shape of the output before" would be better. Second point is to explain it better, since it is unclear how the *final* output shape is computed.

Reply:
I agree this part is tricky. Let me try to explain the idea and you might suggest a better writing for the PR

`im2col`'s `input` returns an `output` which is then used as `input` for `col2im` (this op).
Here,  `col2im`'s `output_shape` must match `im2col`'s `input`.

That is why I mentioned "The shape of the input before applying im2col transformation"